File: python-examples/07-lifecycle-hooks/tools.py
"""Example tools for the lifecycle hooks demo."""

from pydantic import BaseModel
from polos import tool, WorkflowContext


class SearchInput(BaseModel):
    """Input for search tool."""

    query: str


class SearchResult(BaseModel):
    """Output from search tool."""

    results: list[str]
    total_count: int


@tool(description="Search for information on a topic")
async def search(ctx: WorkflowContext, input: SearchInput) -> SearchResult:
    """Simulated search tool for demo purposes."""
    # Simulate search results based on query
    query = input.query.lower()

    results = [
        f"Result 1 for '{input.query}'",
        f"Result 2 for '{input.query}'",
        f"Result 3 for '{input.query}'",
    ]

    return SearchResult(results=results, total_count=len(results))


class CalculatorInput(BaseModel):
    """Input for calculator tool."""

    expression: str


class CalculatorOutput(BaseModel):
    """Output from calculator tool."""

    result: float | None
    error: str | None = None


@tool(description="Calculate a mathematical expression")
async def calculate(ctx: WorkflowContext, input: CalculatorInput) -> CalculatorOutput:
    """Simple calculator tool for demo purposes."""
    try:
        # Only allow safe math operations
        allowed_chars = set("0123456789+-*/().% ")
        if not all(c in allowed_chars for c in input.expression):
            return CalculatorOutput(result=None, error="Invalid characters in expression")

        result = eval(input.expression)  # noqa: S307 - Safe due to character whitelist
        return CalculatorOutput(result=float(result))
    except Exception as e:
        return CalculatorOutput(result=None, error=str(e))

File: typescript-examples/07-lifecycle-hooks/tools.ts
/**
 * Example tools for the lifecycle hooks demo.
 */

import { defineTool, type WorkflowContext } from '@polos/sdk';
import { z } from 'zod';

// ── search ──────────────────────────────────────────────────────────

const searchInputSchema = z.object({
  query: z.string(),
});

const searchOutputSchema = z.object({
  results: z.array(z.string()),
  total_count: z.number(),
});

type SearchInput = z.infer<typeof searchInputSchema>;
type SearchOutput = z.infer<typeof searchOutputSchema>;

export const search = defineTool(
  {
    id: 'search',
    description: 'Search for information on a topic',
    inputSchema: searchInputSchema,
    outputSchema: searchOutputSchema,
  },
  async (_ctx: WorkflowContext, input: SearchInput): Promise<SearchOutput> => {
    // Simulated search results based on query
    const results = [
      `Result 1 for '${input.query}'`,
      `Result 2 for '${input.query}'`,
      `Result 3 for '${input.query}'`,
    ];

    return { results, total_count: results.length };
  },
);

// ── calculate ───────────────────────────────────────────────────────

const calculateInputSchema = z.object({
  expression: z.string(),
});

const calculateOutputSchema = z.object({
  result: z.number().nullable(),
  error: z.string().nullable().optional(),
});

type CalculateInput = z.infer<typeof calculateInputSchema>;
type CalculateOutput = z.infer<typeof calculateOutputSchema>;

export const calculate = defineTool(
  {
    id: 'calculate',
    description: 'Calculate a mathematical expression',
    inputSchema: calculateInputSchema,
    outputSchema: calculateOutputSchema,
  },
  async (_ctx: WorkflowContext, input: CalculateInput): Promise<CalculateOutput> => {
    try {
      // Only allow safe math characters
      const allowed = /^[0-9+\-*/.() %]+$/;
      if (!allowed.test(input.expression)) {
        return { result: null, error: 'Invalid characters in expression' };
      }

      // eslint-disable-next-line no-eval -- safe: validated to only contain math chars
      const result = eval(input.expression) as number;
      return { result: Number(result), error: null };
    } catch (e) {
      return { result: null, error: String(e) };
    }
  },
);
